apps/server/src/modules/tool/uc/tool-reference.uc.ts

```diff
@@ -0,0 +1,100 @@
+import { ForbiddenException, Injectable } from '@nestjs/common';
+import { ToolReference } from '@shared/domain/domainobject/tool/tool-reference';
+import {
+	ContextExternalToolDO,
+	ContextRef,
+	EntityId,
+	ExternalToolDO,
+	Permission,
+	SchoolExternalToolDO,
+	ToolConfigurationStatus,
+} from '@shared/domain';
+import { Action } from '@src/modules/authorization';
+import {
+	CommonToolService,
+	ContextExternalToolService,
+	ExternalToolService,
+	SchoolExternalToolService,
+} from '../service';
+import { ToolContextType } from '../interface';
+import { ToolReferenceMapper } from '../controller/mapper/tool-reference.mapper';
+
+@Injectable()
+export class ToolReferenceUc {
+	constructor(
+		private readonly externalToolService: ExternalToolService,
+		private readonly schoolExternalToolService: SchoolExternalToolService,
+		private readonly contextExternalToolService: ContextExternalToolService,
+		private readonly commonToolService: CommonToolService
+	) {}
+
+	async getToolReferences(userId: EntityId, contextType: ToolContextType, contextId: string): Promise<ToolReference[]> {
+		const contextRef = new ContextRef({ type: contextType, id: contextId });
+
+		const contextExternalTools: ContextExternalToolDO[] = await this.contextExternalToolService.findAllByContext(
+			contextRef
+		);
+
+		const toolReferencesPromises: Promise<ToolReference | null>[] = contextExternalTools.map(
+			(contextExternalTool: ContextExternalToolDO) => this.buildToolReference(userId, contextExternalTool)
+		);
+
+		const toolReferencesWithNull: (ToolReference | null)[] = await Promise.all(toolReferencesPromises);
+		const filteredToolReferences: ToolReference[] = toolReferencesWithNull.filter(
+			(toolReference: ToolReference | null): toolReference is ToolReference => toolReference !== null
+		);
+
+		return filteredToolReferences;
+	}
+
+	private async buildToolReference(
+		userId: EntityId,
+		contextExternalTool: ContextExternalToolDO
+	): Promise<ToolReference | null> {
+		try {
+			await this.ensureToolPermissions(userId, contextExternalTool);
+		} catch (e: unknown) {
+			if (e instanceof ForbiddenException) {
+				return null;
+			}
+		}
+
+		const schoolExternalTool: SchoolExternalToolDO = await this.fetchSchoolExternalTool(contextExternalTool);
+		const externalTool: ExternalToolDO = await this.fetchExternalTool(schoolExternalTool);
+
+		const status: ToolConfigurationStatus = this.commonToolService.determineToolConfigurationStatus(
+			externalTool,
+			schoolExternalTool,
+			contextExternalTool
+		);
+
+		const toolReference: ToolReference = ToolReferenceMapper.mapToToolReference(
+			externalTool,
+			contextExternalTool,
+			status
+		);
+
+		return toolReference;
+	}
+
+	private async ensureToolPermissions(userId: EntityId, contextExternalTool: ContextExternalToolDO): Promise<void> {
+		const promise: Promise<void> = this.contextExternalToolService.ensureContextPermissions(
```

Comment:
Ah i see, ok 2 more things. We speak yesterday over the decision that authorization should only done in UC. After speaking we do *not* change our rule. For tools it should not exists any need to use checkPermissionByReferences it should be use checkPermission and resolve in UC the loading of needed ressources like user and tool. 

If this fixed, then a normal authorizationContext can be one time build and use. 
The method in service can be removed and only the private method in UC stay alive.

Reply:
I saw the deprecation of checkPermissionByReferences, but at the moment we are not able to use checkPermission because it needs a user entity instead of an domain object. So i decided to use checkPermissionByReferences anyway instead of injecting the user repo which would be imo a bigger technical debt. 

This is the uc which does the permission check like the arc group decided. The service just hold a function because multiple uc's use this function because we dont want code duplication and uc's should not call other uc's.